changes/2020-06-04_how-to-fail-with-keyrings/background.md

```diff
@@ -0,0 +1,166 @@
+
+- A mechanism to interrogate a keyring to determine
+  the requirements of its decryption contract
+  (defined in [Issues and Alternatives](#issues-and-alternatives))
+  is out of scope for this document.
+  We should pursue this once decryption contracts are defined
+  for each keyring.
+
+# Issues and Alternatives
+
+## Issue 0 : What does a keyring do?
+
+In order to define when a keyring fails,
+we must first clearly define what a keyring does.
+When a keyring encrypts a data key,
+it creates one or more artifacts that define
+a set of requirements that MUST be met
+in order to use that artifact
+to obtain the data key.
+We call this set of requirements the "decryption contract".
+In order to obtain the data key from the artifact,
+a keyring MUST "satisfy" the decryption contract.
+
+For example, some decryption contracts for keyrings include:
+
+- Raw RSA keyring
+
+  - The keyring MUST have access to the private wrapping key.
+
+- Raw AES keyring
+
+  - The keyring MUST have access to the symmetric wrapping key.
+  - The encryption context MUST exactly match
+    the encryption context used on encrypt.
+
+    - NOTE: In compliant implementations this is a MUST,
+      but mathematically this is a SHOULD
+      because it is possible to create a keyring
+      that ignores this clause in the decryption contract
+      by using AES-CTR rather than AES-GCM.
+
+- AWS KMS keyring (single CMK)
+
+  - The AWS KMS client MUST be configured with credentials for
+    an AWS principal that will pass a `kms:Decrypt` authorization check
+    for the encrypted data key.
+  - The encryption context MUST exactly match
+    the encryption context used on encrypt.
+
+- Multi-keyring:
+
+  - Satisfying any member keyring's decryption contract
+    MUST be sufficient to satisfy
+    the multi-keyring's decryption contract.
+
+## Issue 1 : When MUST a keyring fail?
+
+- On encrypt,
+  if a keyring is unable or unwilling to create an artifact or artifacts
+  that can be used to satisfy its decryption contract,
+  the keyring MUST fail.
+
+- On decrypt,
+  if a keyring is unable or unwilling to satisfy
+  the artifact's decryption contract,
+  the keyring MUST fail.
+
+## Issue 2 : What happens when a keyring fails?
+
+When a keyring fails,
+it MUST communicate
+information about the cause of that failure
+to the component that called the keyring.
+
+## Issue 3 : What implications does this have to existing keyring specifications?
+
+In addition to introducing the concept of decryption contracts
+and defining each keyring's decryption contract,
+this change affects the following keyring specifications.
+
+- Keyring interface :
+  The keyring interface defines no-op behavior as
+  "output[ing] the encryption/decryption materials unmodified".
+  This needs to be modified to frame this behavior as a failure
+  and state that encryption/decryption materials MUST NOT be modified
+  when a keyring fails.
+
+  - _This also affects all keyring specifications._
```

Comment:
I think there are two different failures cases:
1. I didn't see anything which matches my decryption contract and so I didn't even attempt anything. (NOP)
2. I saw something I thought I _could_ decrypt, but failed for some reason.